demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -37,72 +44,197 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
         return eAzureIoTErrorFailed;
     }
 
-    LogInfo( ( "Unencoded the base64 encoding\r\n" ) );
+    AZLogInfo( ( "Unencoded the base64 encoding\r\n" ) );
 
     return eAzureIoTSuccess;
 }
 
 AzureIoTResult_t AzureIoTPlatform_Init( AzureADUImage_t * const pxAduImage )
 {
-    (void) pxAduImage;
-
-    // TODO: Fill in initialization for ST flash writing capability.
+    pxAduImage->xUpdatePartition = ( uint8_t * ) ( FLASH_BASE + FLASH_BANK_SIZE );
+    pxAduImage->ulCurrentOffset = 0;
+    pxAduImage->ulImageFileSize = 0;
+
+    static FLASH_EraseInitTypeDef xEraseInitStruct;
+    uint32_t ulPageError;
+    FLASH_OBProgramInitTypeDef xOptionBytes;
+    AzureIoTResult_t xResult = eAzureIoTSuccess;
+
+    /* Clear OPTVERR bit set on virgin samples. */
+    __HAL_FLASH_CLEAR_FLAG( FLASH_FLAG_OPTVERR );
+    /* Get current optionbytes configuration */
+    HAL_FLASHEx_OBGetConfig( &xOptionBytes );
+
+    /* If BFB2 (Boot From Bank 2) is set, erase bank 1, otherwise erase bank 2 */
+    xEraseInitStruct.Banks =
+        ( ( xOptionBytes.USERConfig & OB_BFB2_ENABLE ) == OB_BFB2_ENABLE )
+        ? FLASH_BANK_1
+        : FLASH_BANK_2;
+    xEraseInitStruct.TypeErase = FLASH_TYPEERASE_MASSERASE;
+
+    HAL_FLASH_Unlock();
+
+    /* Erase non-boot bank */
+    if( HAL_FLASHEx_Erase( &xEraseInitStruct, &ulPageError ) != HAL_OK )
+    {
+        /* Error occurred during page erase. */
+        AZLogError( ( "Error erasing flash bank\r\n" ) );
+        xResult = eAzureIoTErrorFailed;
+    }
 
-    LogInfo( ( "AzureIoTPlatform_Init()\r\n" ) );
+    HAL_FLASH_Lock();
 
-    return eAzureIoTSuccess;
+    return xResult;
 }
 
 AzureIoTResult_t AzureIoTPlatform_WriteBlock( AzureADUImage_t * const pxAduImage,
                                               uint32_t ulOffset,
                                               uint8_t * const pData,
                                               uint32_t ulBlockSize )
 {
-    (void)pxAduImage;
-    (void)ulOffset;
-    (void)pData;
-    (void)ulBlockSize;
+    uint8_t * pucNextWriteAddr = pxAduImage->xUpdatePartition + ulOffset;
+    uint8_t * pucNextReadAddr = pData;
+    uint8_t * pucBlockEndAddr = pxAduImage->xUpdatePartition + ulOffset + ulBlockSize;
+    bool xIsLastBlock = pxAduImage->ulImageFileSize - ulOffset <= ulBlockSize;
```

Comment:
is the `xIsLastBlock` here and the use below redundant with the subsequent check if the `if`?

aka `( pucNextWriteAddr >= ( pucBlockEndAddr - azureiotflashL475_FLASH_ROW_SIZE ) )`

is there a case where `( pucNextWriteAddr >= ( pucBlockEndAddr - azureiotflashL475_FLASH_ROW_SIZE ) )` would be true but `xIsLastBlock` be false?

Reply:
Nope! `xIsLastBlock` checks whether it's the last block of the whole update, and the second check is to see if it's the last chunk we write of the block we get in that call of the function (since the STM can't write a whole block at a time).

For a sample update, Bold is when `xIsLastBlock` is true and italicized is when `( pucNextWriteAddr >= ( pucBlockEndAddr - azureiotflashL475_FLASH_ROW_SIZE ) )` is true. Each Block would be one run of `AzureIoTPlatform_WriteBlock` and each bullet would be one loop of the if statement.

Block 1
- Chunk 1
- _Chunk  2_

Block 2
- Chunk 1
- _Chunk 2_

Block 3
- **Chunk 1**
- **_Chunk 2_**


I'll add some comments in the code!